cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/linkanalysis/PageRank.scala

```diff
@@ -44,15 +44,28 @@ case class PageRankIterationState(pageRank: Array[Double], error: Double, iterat
 class PageRank(graph: DirectedGraph[Node], params: PageRankParams = PageRankParams())
   extends LinkAnalysis[PageRankIterationState](graph, params, "pagerank") {
 
-  private val outboundCount = new Array[Double](graph.maxNodeId + 1)
+  private lazy val outboundCount = new Array[Double](graph.maxNodeId + 1)
+  private val nodeCount = graph.nodeCount
 
-  graph foreach { node => efficientNeighbors(node) foreach { nbr => outboundCount(nbr) += 1} }
+  private val dangling = if (isInStored) {
+    val nonDangling = graph.foldLeft(Set.empty[Int]){ (partialSet, node) =>
+      val neighbors = efficientNeighbors(node)
+      neighbors foreach { nbr => outboundCount(nbr) += 1 }
+      partialSet ++ neighbors.toSet
+    }
+    graph.map { _.id }.toSet -- nonDangling
+  }
+  else {
+    graph.foldLeft(Set.empty[Int]){ (partialSet, node) =>
+      if (efficientNeighbors(node).isEmpty) partialSet + node.id else partialSet
+    }
+  }
```

Comment:
This is correct. outboundCount is 0 because there are no outboundnodes in the graph that only stores the in direction. For a graph which _only_ stores the 'in' direction (i.e., not the out direction as well), we would need to calculate explicitly outboundCount.